### PR TITLE
restore TCP interconnect

### PR DESCRIPTION
**Pull request by pengzhout**
This restore TCP interconnect and fix some hang issues under TCP interconnect.

### CODE REVIEW

File: src/backend/cdb/motion/ic_tcp.c [hunk 1]
```diff
@@ -0,0 +1,2950 @@
+/*-------------------------------------------------------------------------
+ * ic_tcp.c
+ *	   Interconnect code specific to TCP transport.
+ *
+ * Copyright (c) 2005-2008, Greenplum
```
Comment (hsyuan) [inline src/backend/cdb/motion/ic_tcp.c]: copyright

File: src/backend/cdb/motion/ic_tcp.c [hunk 2]
```diff
@@ -0,0 +1,2950 @@
...
+ *
+ * Reviewers: jzhang, tkordas
```
Comment (hsyuan) [inline src/backend/cdb/motion/ic_tcp.c]: do we need this?

File: src/backend/cdb/cdbvars.c [hunk 3]
```diff
@@ -201,6 +208,8 @@ int			interconnect_setup_timeout = 7200;
 
 int			Gp_interconnect_type = INTERCONNECT_TYPE_UDPIFC;
 
+bool		gp_interconnect_aggressive_retry=true; /* fast-track app-level retry */
```
Comment (hsyuan) [inline src/backend/cdb/cdbvars.c]: space around `=`

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 4]
```diff
@@ -950,7 +949,12 @@ static CdbProcess *makeCdbProcess(SegmentDatabaseDescriptor *segdbDesc)
 	}
 
 	process->listenerAddr = pstrdup(qeinfo->hostip);
-	process->listenerPort = segdbDesc->motionListener;
+
+	if (Gp_interconnect_type == INTERCONNECT_TYPE_UDPIFC)
+		process->listenerPort = (segdbDesc->motionListener >> 16) & 0x0ffff;
+	else if (Gp_interconnect_type == INTERCONNECT_TYPE_TCP)
+		process->listenerPort = (segdbDesc->motionListener & 0x0ffff);
```
Comment (hsyuan) [inline src/backend/cdb/dispatcher/cdbgang.c]: maybe we can extract this part as a internal function. my 2 cents

File: src/backend/cdb/motion/ic_tcp.c (hunk 1 above)
Comment (schubert) [inline src/backend/cdb/motion/ic_tcp.c]: Technically copyright should be "Greenplum, Inc.".

**Comment by pengzhout** [comment]
@pivotal-mike thanks for pointing this out, we discussed this yesterday, our strategy is creating a team pipeline to run all jobs of gpdb-master under TCP interconnect, the pipeline will be triggered weekly or monthly, we think it's unnecessary to trigger it for each commit, the TCP and UDP interconnect should make no difference normally and history shows only 4 commits once broken the TCP interconnect in the past year.

**Comment by ashwinstar** [comment]
On Wed, May 31, 2017 at 8:24 PM, Tang Pengzhou <notifications@github.com>
wrote:

> @pivotal-mike <https://github.com/pivotal-mike> thanks for pointing this
> out, we discussed this yesterday, our strategy is creating a team pipeline
> to run all jobs of gpdb-master under TCP interconnect, the pipeline will be
> triggered weekly or monthly, we think it's unnecessary to trigger it for
> each commit, the TCP and UDP interconnect should make no difference
> normally and history shows only 4 commits once broken the TCP interconnect
> in the past year.
>
Is there no way to have a more targeted tests which can be run always
instead of having separate pipelines for smaller-smaller every feature /
customization in product ?

**Comment by pengzhout** [comment]
@ashwinstar An alternative is exposing bellow jobs in main pipeline.
```
 icw_planner_tcp_interconnect_async_dispatcher_centos6, 
 icw_planner_tcp_interconnect_sync_dispatcher_centos6, 
 icw_planner_udpifc_interconnect_async_dispatcher_centos6, 
 icw_planner_udpifc_interconnect_sync_dispatcher_centos6

 icw_gporca_tcp_interconnect_async_dispatcher_centos6, 
 icw_gporca_tcp_interconnect_sync_dispatcher_centos6, 
 icw_gporca_udpifc_interconnect_async_dispatcher_centos6, 
 icw_gporca_udpifc_interconnect_sync_dispatcher_centos6
```
IMO, this may blow up the main pipeline and is unnecessary for each commit, any thought? @yydzero @pivotal-mike

**Comment by pengzhout** [comment]
@ashwinstar @pivotal-mike Added a pipeline job to run icw under TCP interconnect.

**Comment by pivotal-mike** [comment]
That sounds good with only one slight change - I would suggest that you add the job/pipeline to the master pipeline and run it on a trigger.  The reason why I suggest this is so that if something does break we can see it right away instead of trying to remember to check a second pipeline.

@schubert for context

File: concourse/pipelines/pipeline.yml [hunk 5]
```diff
@@ -512,6 +512,25 @@ jobs:
       BLDWRAP_POSTGRES_CONF_ADDONS: "fsync=off"
       TEST_OS: sles
 
+- name: icw_planner_ictcp_centos6
```
Comment (pivotal-mike) [inline concourse/pipelines/pipeline.yml]: Sweet!  Thank you @pengzhout

**Comment by schubert** [comment]
Yes - I'd rather we just run everything on every commit and keep in the main pipeline. We want as little as possible that is "out of sight, out of mind" - coverity being the only clear exception due to the limitations the coverity scanner places on how frequently we can run (twice a week). 

No matter how little things break here, transparency is important.

**Comment by ashwinstar** [comment]
On Thu, Jun 1, 2017 at 1:59 AM, Tang Pengzhou <notifications@github.com>
wrote:

> @ashwinstar <https://github.com/ashwinstar> @pivotal-mike
> <https://github.com/pivotal-mike> Added a pipeline job to run icw under
> TCP interconnect.
>
I understand it not easy answer but will still ask and wish if we can give
one more further thought to find ANY possibility, Is there no other smaller
targeted way of testing this piece than bombarding with full ICW by
enabling the GUC.

Like there may be certain functionality which differs in TCP interconnect
compared to UDP, rest of code should be transparent, if someway we can test
those API's or functions with some targeted minimal tests ?

Just wish to think from perspective if we don't have option of creating
full ICW job, how would one test any change to TCP interconnect.

**Comment by pengzhout** [comment]
@yydzero @ashwinstar thanks, so far we have only one pulse project to test udpifc protocol, we have neither specialized ICW functional tests nor unit tests for both UDPIFC and TCP. In this commit, i can add maybe 4~5 queries that TCP will differ from UDP by experience, but frankly speaking, interconnect highly interacted with executor/dispatcher, i'm not sure if it's enough. For long term, targeted testing is needed, can we add it to the icebox?

**Comment by pengzhout** [comment]
Added a another fix for TCP,  ICW case "external_table" can cover it.

**Comment by pengzhout** [comment]
Merged , thanks for review

**Pull request closed without merging**